Making a kit from an existing package (#93)
MDAnalysis/MDAKits: docs/source/makingakit.rst

Since we are recreating the RMSF analysis,  we are simply copy and pasting the
analysis class into the ``rmsfkit/rmsfkit.py`` file, where the documentation has been

> trimmed for the sake of brevity. Documentation is written in `reStructuredText <https://docutils.sourceforge.io/rst.html>`__

Reviewer: Weird, reST should work with a single underscore (see [sphinx: hyperlinks](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#external-links)). You could make it a named link. 

Have you seen any documentation that states that double-underscore links are supported or even needed?

Although this is the minimum, we highly recommend that you consider generating your documentation with dedicated tools such as `Sphinx <https://www.sphinx-doc.org/en/master/>`_, which allows you to generate static documentation using `reStructuredText <https://www.sphinx-doc.org/en/master/usage/restructuredtext/index.html>`_ formatted plain-text directly from your code.
This makes it easier for your documentation to change alongside code changes.

Testing
*******

We also require that minimal regression tests are present.
These tests are not just useful for when you make changes to your code, but also when using your code with newer versions of MDAnalysis/Python.
These tests are to signify to the users of your packages that the code performs at least the way you say it should and give them confidence that it can be used.
Basic tests can be written with a variety of packages, such as the `unittest package <https://docs.python.org/3/library/unittest.html#module-unittest>`_ or the `pytest package <https://docs.pytest.org/en/7.4.x/>`_.

> Futher improvements to your testing proceedure may include automatically running the tests on pushing to your remote repositories (see `GitHub Actions <https://github.com/features/actions>`_).

Reviewer: I'd like to see here (and elsewhere) some concrete examples, i.e., the "typical" code that would go into the metadata.yml file.

.. image:: img/rmsftutorial/creating_a_release.gif
	:alt: Process for creating a release by making a tag

Building from an existing project
#################################

Registering an existing package as an MDAKit is a straightforward process.
Since the structure of an MDAKit is not as strict as the code found in the MDAnalysis core library, chances are very little restructuring is needed for registration.

> The primary concern is ensuring that the core MDAKit requirements are met, as listed at the top of this document.

Reviewer: link so that one can jump to the requirements

Licensing
*********

One of the more pressing requirements for kit registration is clearly identifying the license that is applied to your code.
This is typically included in a LICENSE file at the top level of your repository.
`Without a license <https://choosealicense.com/no-permission/>`_, the only assumption a user can make about your code is that they are not in a position to use your code.
Your license needs to be compatible with the GPLv2+ license currently used by MDAnalysis.

> Take time to consider how you would like to license your project.

Reviewer: add a link to choosealicense to help them

Hosting code in a version controlled repository
***********************************************

Since the MDAKits registry makes heavy use of the GitHub actions infrastructure, registration of a kit requires that all code maintainers also have a GitHub account for communication purposes.

> For this reason, if your code is not already hosted in an accessible version controlled repository, hosting on `GitHub <https://github.com>`_ is recommended, although other services such as `BitBucket <https://bitbucket.org/>`_, `GitLab <https://gitlab.com>`_, or self hosting is possible.

Reviewer: add here explanation of the `source_install` entry, see #98

Although this is the minimum, we highly recommend that you consider generating your documentation with dedicated tools such as `Sphinx <https://www.sphinx-doc.org/en/master/>`_, which allows you to generate static documentation using `reStructuredText <https://www.sphinx-doc.org/en/master/usage/restructuredtext/index.html>`_ formatted plain-text directly from your code.
This makes it easier for your documentation to change alongside code changes.

Testing
*******

We also require that minimal regression tests are present.
These tests are not just useful for when you make changes to your code, but also when using your code with newer versions of MDAnalysis/Python.
These tests are to signify to the users of your packages that the code performs at least the way you say it should and give them confidence that it can be used.
Basic tests can be written with a variety of packages, such as the `unittest package <https://docs.python.org/3/library/unittest.html#module-unittest>`_ or the `pytest package <https://docs.pytest.org/en/7.4.x/>`_.

> Futher improvements to your testing proceedure may include automatically running the tests on pushing to your remote repositories (see `GitHub Actions <https://github.com/features/actions>`_).

Reviewer: Use the word "continuous integration (CI)" somewhere so that people know what we're talking about when we mention it. The tone and level of the existing text is good, though.

> When submitting an MDAKit to the registry, you can include the instructions for running the tests in the required ``metadata.yaml`` file (see a full example in the `registration <registration>`__ section below).

Reviewer: Not "can" but "do", see #95

.. image:: img/rmsftutorial/creating_a_release.gif
	:alt: Process for creating a release by making a tag

Building from an existing project
#################################

Registering an existing package as an MDAKit is a straightforward process.
Since the structure of an MDAKit is not as strict as the code found in the MDAnalysis core library, chances are very little restructuring is needed for registration.

> The primary concern is ensuring that the core MDAKit :ref:`requirements<requirements>` are met, as listed at the top of this document.

Reviewer: This ref works; just IIRC the sphinx docs say somewhere "space between text and link name", i.e., `:ref: requirements  <requirements>`. Come to think, if title and link name are the same, then just
```reST
:ref:`requirements`
```
_should_ work. (But test!)

Licensing
*********

One of the more pressing requirements for kit registration is clearly identifying the license that is applied to your code.
This is typically included in a LICENSE file at the top level of your repository.
`Without a license <https://choosealicense.com/no-permission/>`_, the only assumption a user can make about your code is that they are not in a position to use your code.

> Your license needs to be compatible with the GPLv2+ license currently used by MDAnalysis.

Reviewer 2: ```suggestion
Your license needs to be compatible with the GPLv2+ license currently used by MDAnalysis, in addition to the licenses of any other packages your mdakit depends on.
```
Maybe not the best worded suggestion - essentially "don't just look at MDA" (this is the indirect "addressing Apache 2.0" comment)

> Take time to consider how you would like to `license your project <https://choosealicense.com/>`_.

Reviewer 2: I've always found choose a license to be a bit opiniated in the way it thinks about licenses.

How about just throwing a couple of resources here to avoid us endorsing anything specific?

```suggestion
Take time to consider how you would like to license your project. Further information on open source licensing can be found from sources such as;  `choose a license <https://choosealicense.com/>`_, `tl;dr Legal <https://tldrlegal.com/>`_, the `Open Source Initiative <https://opensource.org/osd>`_, and the `Software Sustainability Insitute <https://www.software.ac.uk/resources/guides/choosing-open-source-licence>`_.
```

Hosting code in a version controlled repository
***********************************************

Since the MDAKits registry makes heavy use of the GitHub actions infrastructure, registration of a kit requires that all code maintainers also have a GitHub account for communication purposes.
For this reason, if your code is not already hosted in an accessible version controlled repository, hosting on `GitHub <https://github.com>`_ is recommended, although other services such as `Bitbucket <https://bitbucket.org/>`_, `GitLab <https://gitlab.com>`_, or self hosting is possible.

> With publicly accessible code, users are not required to get go through intermediate packaging services such as the Python Package Index or conda-forge, although having your code available through these services is highly encouraged.

Reviewer: "to get go through" — ?

omit "get"??

> With publicly accessible code, users are not required to get go through intermediate packaging services such as the Python Package Index or conda-forge, although having your code available through these services is highly encouraged.

Reviewer 2: ```suggestion
The registry does not require that your code be available through packaging repositories such as the Python Package Index or conda-forge, although having your code available through these services is highly encouraged.
```
Slight rewording, please ignore if you don't like it - "intermediate packaging services" felt like a bit on the technical side of things.

After registration, users can find the installation instructions for the source code on your MDAKit page, which is specified in the ``src_install`` field in the ``metadata.yaml`` file (see :ref:`specification`).

Documentation
*************

Basic documentation is required for MDAKit registration.
The detail and depth of the documentation is ultimately up to you, but we require at a minimum that you provide README-style documentation explaining what the code is supposed to do, how to install it, and the basics of its use.
Although this is the minimum, we highly recommend that you consider generating your documentation with dedicated tools such as `Sphinx <https://www.sphinx-doc.org/en/master/>`_, which allows you to generate static documentation using `reStructuredText <https://www.sphinx-doc.org/en/master/usage/restructuredtext/index.html>`_ formatted plain-text directly from your code.
This makes it easier for your documentation to change alongside code changes.

> Using a documentation hosting service such as `Read the Docs <https://readthedocs.org>`_ or `GitHub Pages <https://pages.github.com/>`_ makes public access to your generated documentation automatic.

Reviewer 2: @fiona-naughton these sections might be good entry points for further guides, i.e. this can eventually be switched to "see our guide on building docs for mdakits here".

Testing
*******

We also require that minimal regression tests are present.

> These tests are not just useful for when you make changes to your code, but also when using your code with newer versions of MDAnalysis/Python.

Reviewer 2: ```suggestion
These tests are not just useful for when you make changes to your code, but also when any package dependencies (e.g. MDAnalysis, NumPy, and Python) change.
```

These tests are to signify to the users of your packages that the code performs at least the way you say it should and give them confidence that it can be used.

> Basic tests can be written with a variety of packages, such as the `unittest package <https://docs.python.org/3/library/unittest.html#module-unittest>`_ or the `pytest package <https://docs.pytest.org/en/7.4.x/>`_.

Reviewer 2: Should we say that we recommend the latter?

Reviewer: I would want to write 
> "If you don't have settled on a testing framework yet then choose pytest but if you use something else already then that's totally cool because the most important thing is that you _have_ tests." 

but that's very wordy and might just distract.

I am happy to leave it neutral but if one of you has a strong opinion here then we can put it in.

Author: I don't have a strong opinion either way. Whichever framework speaks to the authors is the one they should choose as far as I'm concerned.

Further improvements to your testing procedure may include automatically running the tests on pushing to your remote repositories, often referred to as continuous integration (CI).
CI can be set up using repository pipeline tools, such as `GitHub Actions <https://github.com/features/actions>`_.

When submitting an MDAKit to the registry, include the instructions for running the tests in the required ``metadata.yaml`` file (see a full example in the `registration <registration_>`_ section below).
Assuming that your tests are in a ``test/`` directory at the top level of your repository, you could define your test commands as:

.. code-block:: yaml

	run_tests:

> 	  - git clone https://github.com/MYGITHUB/MYPACKAGE --depth=1

Reviewer 2: I'm going to put a change request here - this should either use "git clone latest" or "pip/conda install".

Reviewer 2: ### Why not a regular git clone command?

This will break when running "latest" tests if things have change on the head branch of your repo. We've seen issues with this in practice.

### Should we just encourage pip / conda install instead?

Overall getting folks to package their tests would be easier - git clone latest is a not so amazing workaround (it makes assumptions on using the latest version tag). That being said we'd then need to tell folks about packaging their tests, which is a headache in its own rights.

	  - cd MYPACKAGE

> 	  - pytest -v tests/

Reviewer: As written here, this is a safe bet, where everything is explicit.

The docs in the meta.yml file indicate that one can short-cut:
```yaml
# For example, if the tests are in the repository at the top level under `./tests`:
## First use `git clone latest` to either clone the top commit for "development code" checks or check out
## the latest tag for "latest release" checks. Then then run pytest:
##    - git clone latest
##    - pytest -v ./tests
```

What are the pre-requisites for `git clone latest` working properly, @IAlibay ? Does it depend on `src_install` being defined? (I think this is effectively related to #98 )

For this PR I am ok with having the generic instructions. Later (as part of #98 possibly) we should rethink the docs.

Reviewer 2: `git clone latest` depends on `project_home` (which the more I think probably should have been called `src_home`).
